### PR TITLE
sql: Prevent dropping indexes and tables referenced by FKs

### PR DESCRIPTION
**Pull request by dt**
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6832)

### CODE REVIEW

File: sql/drop.go [hunk 1]
```diff
@@ -211,6 +211,21 @@ func (n *dropIndexNode) Start() error {
 		// Queue the mutation.
 		switch status {
 		case sqlbase.DescriptorActive:
+			idx := tableDesc.Indexes[i]
+
+			// If an index is either end of a FK constraint, it cannot be dropped.
```
Comment (bdarnell) [inline sql/drop.go]: If `n.n.DropBehavior == parser.DropCascade`, then the FK should be dropped as well to allow the `DROP INDEX` to proceed. We don't need to implement this now, but we might want to use a different error message to acknowledge that this is not yet supported. (Some frameworks add `CASCADE` to all their `DROP` commands so we want to accept that whenever there is nothing to actually cascade to).

Comment (dt) [inline sql/drop.go]: Updated with explicit mention of CASCADE being unsupported for now.

**Comment by dt** [comment]
Cc @paperstreet

**Comment by vivekmenezes** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/drop.go, line 310 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_g0sJMMuWMBG2PtiS:-KI_g0sJMMuWMBG2PtiT:-1646094749) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/drop.go#L310)):*
> ```Go
> 				}
> 				return nil, fmt.Errorf("%q is referenced by foreign key from table %q", droppedDesc.Name, t.Name)
> 			}
> ```

It appears that this code is a copy of the code change you've made above. Perhaps you can abstract it into a new method.

---

*[sql/drop.go, line 431 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_hxhx2eRjuV-WSDwv:-KI_hxhx2eRjuV-WSDww:528691006) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/drop.go#L431)):*
> ```Go
> 			p.notifySchemaChange(t.ID, sqlbase.InvalidMutationID)
> 		}
> 	}
> ```

I think you need to first modify the other tables and ensure that there are no leases on those tables before you can delete the original table. To do that correctly you need to move this code to where the table is being deleted.

So the sequence is
1. mark the table as dropped and set upversion

in the schema changer:

2. if the table is marked as deleted, remove all references of that table from other tables and setupversion for all the tables with references.

3. Run the schema changers for all the tables with references to ensure there are no outstanding leases on descriptors with references

4. delete the original table.

---

*[sql/testdata/fk, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_jTVSnKEaHob-eZRS:-KI_jTVSnKEaHob-eZRT:1915370709) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/testdata/fk#L97)):*
> ```
> 
> statement ok
> DROP TABLE products
> ```

Just ensure you're testing tables with more than one inbound and outbound reference. Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832)*

**Comment by dt** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sql/drop.go, line 310 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_g0sJMMuWMBG2PtiS:-KIcfivCPgJAk_0zSQNJ:-654839514) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/drop.go#L310)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> It appears that this code is a copy of the code change you've made above. Perhaps you can abstract it into a new method.
</details>
Done.

It wasn't quite the same code though, since it included the kind of thing being dropped and its name, thus a reusable helper has to be relatively parameterized and, amusingly, actually nets out to be more code than the inline version was.

---

*[sql/testdata/fk, line 97 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_jTVSnKEaHob-eZRS:-KIcQ1i2ISsWpR84rMpZ:-2039732054) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/testdata/fk#L97)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Just ensure you're testing tables with more than one inbound and outbound reference. Thanks
</details>
ah, yeah, had multiple outbound on `orders` but  added another table to get multiple inbound on `products` plus both directions on `orders`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832)*

**Comment by vivekmenezes** [comment]
:lgtm:

<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary>
> Cc @paperstreet
</details>


---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-:-KId8wQcqNMJYfC9WXQQ:1417088721)*

**Comment by dt** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[sql/drop.go, line 431 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832#-KI_hxhx2eRjuV-WSDwv:-KIdETdMhnK0DzYnne9q:-1277068382) ([raw file](https://github.com/cockroachdb/cockroach/blob/242647142ad0fb935bf713d4f2deefdb2c8ee69a/sql/drop.go#L431)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I think you need to first modify the other tables and ensure that there are no leases on those tables before you can delete the original table. To do that correctly you need to move this code to where the table is being deleted.
> 
> So the sequence is
> 1. mark the table as dropped and set upversion
> 
> in the schema changer:
> 
> 2. if the table is marked as deleted, remove all references of that table from other tables and setupversion for all the tables with references.
> 
> 3. Run the schema changers for all the tables with references to ensure there are no outstanding leases on descriptors with references
> 
> 4. delete the original table.
> 
> 
</details>
Filed #6900 to explicitly track the schema-change tech debt here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6832)*

**Pull request merged**